src/back-end/lib/resources/affiliation.ts

```diff
@@ -58,63 +56,58 @@ const resource: Resource = {
-        } else {
-          // If existing, active affiliation, create new affiliation with ACTIVE status and updated role
-          if (existingAffiliation.membershipStatus === MembershipStatus.Active) {
-            if (!permissions.updateAffiliation(connection, request.session, organization)) {
+        const validatedUser = await validateUserId(connection, user);
+        const validatedOrganization = await validateOrganizationId(connection, organization);
+        const validatedMembershipType = affiliationValidation.validateMembershipType(membershipType);
+        if (allValid([validatedUser, validatedOrganization, validatedMembershipType])) {
+          const existingAffiliation = await readOneAffiliation(connection, user, organization);
+          if (!existingAffiliation) {
+            if (!permissions.createAffiliation(request.session, user)) {
               return invalid({
                 permissions: [permissions.ERROR_MESSAGE]
               });
             }
-
-            const validatedMembershipType = validateMembershipType(membershipType);
-            if (validatedMembershipType.tag === 'invalid') {
+            // If no existing, active affiliation, create new affiliation with PENDING status
+            return valid({
+              user,
+              organization,
+              membershipType: (validatedMembershipType.value as MembershipType),
+              membershipStatus: MembershipStatus.Pending
+            });
+          } else {
+            if (!permissions.updateAffiliation(connection, request.session, organization)) {
               return invalid({
-                membershipType: ['Invalid membership type provided.']
+                permissions: [permissions.ERROR_MESSAGE]
               });
             }
-            return valid({ ...validResponseObject, membershipType, membershipStatus: MembershipStatus.Active });
+            // If existing, active affiliation, create new affiliation with ACTIVE status and updated role
+            return valid({
+              user,
+              organization,
+              membershipType: (validatedMembershipType.value as MembershipType),
+              membershipStatus: MembershipStatus.Active
+            });
           }
+        } else {
+          return invalid({
+            user: getInvalidValue(validatedUser, undefined),
+            organization: getInvalidValue(validatedOrganization, undefined),
+            membershipType: getInvalidValue(validatedMembershipType, undefined)
+          });
         }
-
-        return invalid({
-          membershipType: ['Invalid membership type specified']
-        });
       },
-      async respond(request) {
-        const respond = (code: number, body: Affiliation | string[]) => basicResponse(code, request.session, makeJsonResponseBody(body));
+      async respond(request): Promise<Response<JsonResponseBody<Affiliation | CreateValidationErrors>, Session>> {
```

Comment:
This type signature is good because the `Resource` type only types the response body as one of `SupportedResponseBodies` which is very vague.